public/app/core/components/connectWithCleanUp.tsx

```diff
@@ -0,0 +1,36 @@
+import { MapStateToPropsParam, MapDispatchToPropsParam, connect, useDispatch } from 'react-redux';
+import { StateSelector, cleanUpAction } from '../actions/cleanUp';
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import hoistNonReactStatics from 'hoist-non-react-statics';
+
+export const connectWithCleanUp = <
+  TStateProps extends {} = {},
+  TDispatchProps = {},
+  TOwnProps = {},
+  State = {},
+  TSelector extends object = {}
+>(
+  mapStateToProps: MapStateToPropsParam<TStateProps, TOwnProps, State>,
+  mapDispatchToProps: MapDispatchToPropsParam<TDispatchProps, TOwnProps>,
+  stateSelector: StateSelector<TSelector>
+) => (Component: ComponentType<any>) => {
+  const ConnectedComponent = connect(
+    mapStateToProps,
+    mapDispatchToProps
+  )(Component);
+
+  const ConnectedComponentWithCleanUp: FunctionComponent = props => {
+    const dispatch = useDispatch();
+    useEffect(() => {
+      return () => dispatch(cleanUpAction({ stateSelector }));
+    });
+    // @ts-ignore
+    return <ConnectedComponent {...props} />;
+  };
+
+  ConnectedComponentWithCleanUp.displayName = `ConnectWithCleanUp(${ConnectedComponent.displayName})`;
+  hoistNonReactStatics(ConnectedComponentWithCleanUp, Component);
+  type Hoisted = typeof ConnectedComponentWithCleanUp;
```

Comment:
Is this enough? `ConnectedComponentWithCleanUp` does not actually have the statics so I think nor will the `Hoisted` have. In https://github.com/grafana/grafana/blob/359404eb778f5d5a17915a0687e8598b65cdde9d/packages/grafana-ui/src/themes/ThemeContext.tsx#L21 I had to add extra generic parameter with which a caller can manually specify the hosted statics. Did not find easier way to type that properly

Reply:
Can you give an example where this doesn't work, I'm not 100% sure about the implementation here so I could use some pointers.

Reply:
I moved some constants [here](https://github.com/grafana/grafana/blob/359404eb778f5d5a17915a0687e8598b65cdde9d/packages/grafana-ui/src/components/RefreshPicker/RefreshPicker.tsx#L44) to make it easier to import. When I wrapped the refresh picker in `withTheme` the statics were lost in the typing. Seems like `hoistNonReactStatic` does not change the type properly. Thats why I added the `type Hoisted = typeof WithTheme & S;` so that you could manually specify the statics when using `withTheme` that will be also in the hoisted type.